back-end/controllers/record-controller.js

```diff
@@ -36,38 +59,90 @@ const recordController = {
   // Edit the record
   // URL: put /records/:rid
   putRecord: async (req, res, next) => {
-    try {
-      const recordId = Number(req.params.rid)
-      const { title, amount, note, categoryId } = req.body
-      const userId = req.user.id
+    const recordId = Number(req.params.rid)
+    const { title, amount, note, categoryId } = req.body
+    const userId = req.user.id
+
+    // Check if there is record id in req.params
+    if (!recordId) {
+      return res.status(400).json({
+        type: 'Put record failed',
+        title: 'Require reocrd id',
+        field_errors: {
+          recordId: 'required',
+        },
+      })
+    }
+
+    // Check if there are title, amount and categoryId in req.body
+    if (!title || !amount || !categoryId) {
+      return res.status(400).json({
+        type: 'Put record failed',
+        title: 'Require reocrd title, amount or categoryId',
+        field_errors: {
+          title: 'required',
+          amount: 'required',
+          categoryId: 'required',
+        },
+      })
+    }
 
-      // Check if there are missing data
-      if (!title || !amount || !categoryId)
-        return res
-          .status(400)
-          .json('missing title or amount or category id to update this record')
+    // Check data type
+    if (
+      typeof recordId !== 'number' ||
+      typeof title !== 'string' ||
+      typeof amount !== 'number' ||
+      typeof categoryId !== 'number'
+    ) {
+      return res.status(400).json({
+        type: 'Put record failed',
+        title: 'Incorrect data type',
+        field_errors: {
+          recordId: 'number',
+          title: 'string',
+          amount: 'number',
+          categoryId: 'number',
+        },
+      })
+    }
 
+    try {
       // Check if the record is in database
       const theRecord = await prisma.record.findUnique({
         where: {
           id: recordId,
         },
       })
-      if (!theRecord) return res.status(400).json('the record does not exist')
+
+      // !theRecord: {}
+      if (theRecord === 'null') {
```

Comment:
It seems you want to validate if `theRecord` is an empty object.
If that's the case, I'd suggest using a simpler way like `if (Object.keys(theRecord).length !== 0)`.
It's more expressive and easier for others to understand that you're validating an object.

Reply:
Amazing, You point out the questions I'm thinking about!
Yes, I want to validate if `theRecord` is an empty object, because an empty object would make the following validation process error. I have thought about using `length` method to check the object, but didn't figure out how to practice.

Thank you for teaching me how to practice the checking process.